### ISSUE TITLE
Persisting plugins on the docker host 

### ISSUE THREAD

**Issue by pecuchet** [ether/etherpad-lite#3582]
Hello,
I'm running etherpad-lite through docker (`17.05.0-ce, build 89658b`). It works well, but when I restart the container, plugins which were installed through the admin interface are lost.  

I thought it would only require bind-mounting the plugin folders to the docker host, but the symlink (`ep_etherpad-lite -> ../src`) located in `/opt/etherpad-lite/node_modules` does not seem to survive the mounting &mdash; even if the file paths of the container are mimicked on the host like below. 

Any ideas on how to fix this? 

```
version: '3.4'

services:
  etherpad:
    build: .
    container_name: etherpad
    networks:
      - my-network
    volumes:
      - ./settings.json:/opt/etherpad-lite/settings.json:ro
      - /opt/etherpad-lite/var:/opt/etherpad-lite/var
      - /opt/etherpad-lite/node_modules:/opt/etherpad-lite/node_modules
      - /opt/etherpad-lite/src:/opt/etherpad-lite/src
    environment:
      - NODE_ENV=production

networks:
  my-network:
    external:
      name: ${DOCKER_NETWORK}
```

The container throws the following error upon startup:
```
internal/modules/cjs/loader.js:651
    throw err;
    ^

Error: Cannot find module '/opt/etherpad-lite/node_modules/ep_etherpad-lite/node/server.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:649:15)
    at Function.Module._load (internal/modules/cjs/loader.js:575:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:862:12)
    at internal/main/run_main_module.js:21:11
```

**Comment by muxator** [comment]
Hi @pecuchet, this has to be solved reorganizing the on-disk layout of Etherpad files. I am thinking very insistently about it lately.

Since I think there are no clean solutions right now, a possible **very dirty** "ops-level" tactic could be something along these lines (UNSECURE, UNSUPPORTED, UNTESTED):

- keep the `/opt/etherpad-lite/node_modules` volume. Once you'll have done this dirty trick it will store only the installed plugins
- run the container in `--privileged` mode, to allow the use of the `mount` command inside it. BEWARE: **this has security implications**. Do not do it unless you know what you are doing
- choose your favorite way to modify the startup command, in order to execute this command before starting:
   ```bash
   # I am assuming mounting the volume nuked this symlink.
   # Let's create an empty directory to place a bind mount over it.
   mkdir /opt/etherpad-lite/node_modules/ep_etherpad-lite

   # this is a bind mount performed from inside the container OVER the docker volume
   mount --bind /opt/etherpad-lite/src /opt/etherpad-lite/node_modules/ep_etherpad-lite
   ```

If this works, feel ashamed and do not forget to purify your soul invoking your preferred mystical rites.

Then - please - pitifully drop a line with some documentation in this ticket to help the poor souls of the future speeding their way to an unforgiving hell.

**Comment by muxator** [comment]
Opened #3593 for tracking the technical aspects of moving the plugins to a dedicated directory.

Leaving this open for a while (~days) in case @pecuchet had done some experiments and is willing to share a feedback on the _specific_ problem.

**Comment by pecuchet** [comment]
Hi @muxator, I ended up choosing a more righteous path and steering my soul clear of devilish security torments... This is simply what I did:
```
FROM etherpad/etherpad:latest
RUN npm i ep_mypads
```
Of course it doesn't solve the problem of plugins installed at run-time. But it could be an idea to give the possibility to install plugins through a built-time argument (`docker build --build-arg` and a default Dockerfile `ARG`). This way, all what pertains to the software resides in the image, which adheres to Docker's concept of images and containers.
